File: backend/papertrading/holding_t_signal_test.go
package papertrading_test

import (
	"fmt"
	"testing"
	"time"

	"github.com/stretchr/testify/require"

	"go-stock/backend/papertrading"
)

func tBarsNearCostPullbackBounce() []papertrading.HoldingTSignalBar {
	// Local high 10.20, drift down near cost 10.00, then bounce.
	prices := []struct{ h, l, c float64 }{
		{10.10, 10.00, 10.05},
		{10.15, 10.05, 10.12},
		{10.20, 10.10, 10.18},
		{10.18, 10.08, 10.10},
		{10.12, 10.02, 10.04},
		{10.08, 9.98, 10.00},
		{10.05, 9.97, 9.99},
		{10.06, 9.98, 10.03}, // bounce, near cost
	}
	bars := make([]papertrading.HoldingTSignalBar, len(prices))
	for i, p := range prices {
		bars[i] = papertrading.HoldingTSignalBar{
			Time:  fmt.Sprintf("10:%02d", i*5),
			High:  p.h,
			Low:   p.l,
			Close: p.c,
		}
	}
	return bars
}

func tBarsRiseAwayFade() []papertrading.HoldingTSignalBar {
	// Rise from ~10 to ~10.28 then fade last bar.
	prices := []struct{ h, l, c float64 }{
		{10.05, 9.95, 10.00},
		{10.10, 10.00, 10.08},
		{10.15, 10.05, 10.12},
		{10.20, 10.10, 10.18},
		{10.25, 10.15, 10.22},
		{10.28, 10.18, 10.26},
		{10.30, 10.20, 10.28},
		{10.29, 10.18, 10.20}, // fade
	}
	bars := make([]papertrading.HoldingTSignalBar, len(prices))
	for i, p := range prices {
		bars[i] = papertrading.HoldingTSignalBar{
			Time:  fmt.Sprintf("11:%02d", i*5),
			High:  p.h,
			Low:   p.l,
			Close: p.c,
		}
	}
	return bars
}

func TestBuildHoldingTSignal_NormalBuyWatch(t *testing.T) {
	asOf := time.Date(2026, 9, 8, 10, 35, 0, 0, time.Local)
	res := papertrading.BuildHoldingTSignal(papertrading.HoldingTSignalInput{
		StockCode:        "sh600000",
		PositionID:       "42",
		HasPosition:      true,
		CanSell:          true,
		AvailableQty:     1000,
		CostPrice:        10.0,
		Freshness:        papertrading.EvalFreshnessFresh,
		SuitabilityLevel: papertrading.TSuitabilitySuitable,
		HealthGrade:      papertrading.HealthGradeB,
		Bars:             tBarsNearCostPullbackBounce(),
		AsOf:             asOf,
	})
	require.Empty(t, res.Notes)
	var buy *papertrading.HoldingTSignal
	for i := range res.Signals {
		if res.Signals[i].SignalType == papertrading.TSignalBuyWatch {
			buy = &res.Signals[i]
		}
	}
	require.NotNil(t, buy, "expected T_BUY_WATCH, got %+v", res.Signals)
	require.Equal(t, "sh600000", buy.StockCode)
	require.Equal(t, "42", buy.PositionID)
	require.InDelta(t, 10.03, buy.SignalPrice, 1e-9)
	require.Contains(t, buy.Reasons, papertrading.TSignalReasonNearCost)
	require.Greater(t, buy.Confidence, 0.0)
	require.LessOrEqual(t, buy.Confidence, 1.0)
}

func TestBuildHoldingTSignal_NormalSellWatch(t *testing.T) {
	res := papertrading.BuildHoldingTSignal(papertrading.HoldingTSignalInput{
		StockCode:        "sz000001",
		HasPosition:      true,
		CanSell:          true,
		AvailableQty:     500,
		CostPrice:        10.0,
		Freshness:        papertrading.EvalFreshnessFresh,
		SuitabilityLevel: papertrading.TSuitabilitySuitable,
		HealthGrade:      papertrading.HealthGradeA,
		Bars:             tBarsRiseAwayFade(),
		AsOf:             time.Now(),
	})
	var sell *papertrading.HoldingTSignal
	for i := range res.Signals {
		if res.Signals[i].SignalType == papertrading.TSignalSellWatch {
			sell = &res.Signals[i]
		}
	}
	require.NotNil(t, sell, "expected T_SELL_WATCH, got %+v notes=%v", res.Signals, res.Notes)
	require.Contains(t, sell.Reasons, papertrading.TSignalReasonAwayCost)
	require.Contains(t, sell.Reasons, papertrading.TSignalReasonMomentumFade)
}

func TestBuildHoldingTSignal_CannotSell_NoBuy(t *testing.T) {
	res := papertrading.BuildHoldingTSignal(papertrading.HoldingTSignalInput{
		StockCode:    "sh600000",
		HasPosition:  true,
		CanSell:      false,
		AvailableQty: 0,
		CostPrice:    10.0,
		Freshness:    papertrading.EvalFreshnessFresh,
		Bars:         tBarsNearCostPullbackBounce(),
	})
	require.Contains(t, res.Notes, papertrading.TSignalReasonCannotSell)
	for _, s := range res.Signals {
		require.NotEqual(t, papertrading.TSignalBuyWatch, s.SignalType)
	}
}

func TestBuildHoldingTSignal_PriceStale(t *testing.T) {
	res := papertrading.BuildHoldingTSignal(papertrading.HoldingTSignalInput{
		StockCode:    "sh600000",
		HasPosition:  true,
		CanSell:      true,
		AvailableQty: 100,
		CostPrice:    10.0,
		Freshness:    papertrading.EvalFreshnessStale,
		Bars:         tBarsNearCostPullbackBounce(),
	})
	require.Empty(t, res.Signals)
	require.Contains(t, res.Notes, papertrading.TSignalReasonPriceStale)
}

func TestBuildHoldingTSignal_No5mBars(t *testing.T) {
	res := papertrading.BuildHoldingTSignal(papertrading.HoldingTSignalInput{
		StockCode:    "sh600000",
		HasPosition:  true,
		CanSell:      true,
		AvailableQty: 100,
		CostPrice:    10.0,
		Freshness:    papertrading.EvalFreshnessFresh,
		Bars:         []papertrading.HoldingTSignalBar{{Close: 10, High: 10.1, Low: 9.9}},
	})
	require.Empty(t, res.Signals)
	require.Contains(t, res.Notes, papertrading.TSignalReasonNoBars)
}

func TestBuildHoldingTSignal_EmptyPosition(t *testing.T) {
	res := papertrading.BuildHoldingTSignal(papertrading.HoldingTSignalInput{
		StockCode:   "sh600000",
		HasPosition: false,
		CanSell:     false,
		CostPrice:   10.0,
		Freshness:   papertrading.EvalFreshnessFresh,
		Bars:        tBarsNearCostPullbackBounce(),
	})
	require.Empty(t, res.Signals)
	require.Contains(t, res.Notes, papertrading.TSignalReasonNoPosition)
}

func TestBuildHoldingTSignal_NilSafeZero(t *testing.T) {
	res := papertrading.BuildHoldingTSignal(papertrading.HoldingTSignalInput{})
	require.Empty(t, res.Signals)
	require.Contains(t, res.Notes, papertrading.TSignalReasonNoPosition)
}

File: frontend/src/utils/holdingTSignal.test.mjs
/**
 * Minimal unit checks for holdingTSignal (node --test / vite-node style).
 */
import assert from 'node:assert/strict'
import {
  buildHoldingTSignal,
  holdingTSignalsToChartMarkers,
  T_SIGNAL_BUY_WATCH,
  T_SIGNAL_SELL_WATCH,
  T_SIGNAL_REASON,
} from './holdingTSignal.js'

function barsPullback() {
  return [
    { time: '10:00', high: 10.1, low: 10.0, close: 10.05 },
    { time: '10:05', high: 10.15, low: 10.05, close: 10.12 },
    { time: '10:10', high: 10.2, low: 10.1, close: 10.18 },
    { time: '10:15', high: 10.18, low: 10.08, close: 10.1 },
    { time: '10:20', high: 10.12, low: 10.02, close: 10.04 },
    { time: '10:25', high: 10.08, low: 9.98, close: 10.0 },
    { time: '10:30', high: 10.05, low: 9.97, close: 9.99 },
    { time: '10:35', high: 10.06, low: 9.98, close: 10.03 },
  ]
}

function barsRiseFade() {
  return [
    { time: '11:00', high: 10.05, low: 9.95, close: 10.0 },
    { time: '11:05', high: 10.1, low: 10.0, close: 10.08 },
    { time: '11:10', high: 10.15, low: 10.05, close: 10.12 },
    { time: '11:15', high: 10.2, low: 10.1, close: 10.18 },
    { time: '11:20', high: 10.25, low: 10.15, close: 10.22 },
    { time: '11:25', high: 10.28, low: 10.18, close: 10.26 },
    { time: '11:30', high: 10.3, low: 10.2, close: 10.28 },
    { time: '11:35', high: 10.29, low: 10.18, close: 10.2 },
  ]
}

{
  const res = buildHoldingTSignal({
    stockCode: 'sh600000',
    hasPosition: true,
    canSell: true,
    availableQty: 1000,
    costPrice: 10,
    freshness: 'FRESH',
    suitabilityLevel: 'suitable',
    healthGrade: 'B',
    bars: barsPullback(),
  })
  assert.ok(res.signals.some((s) => s.signal_type === T_SIGNAL_BUY_WATCH))
}

{
  const res = buildHoldingTSignal({
    stockCode: 'sz000001',
    hasPosition: true,
    canSell: true,
    availableQty: 500,
    costPrice: 10,
    freshness: 'FRESH',
    bars: barsRiseFade(),
  })
  assert.ok(res.signals.some((s) => s.signal_type === T_SIGNAL_SELL_WATCH))
}

{
  const res = buildHoldingTSignal({
    hasPosition: true,
    canSell: false,
    availableQty: 0,
    costPrice: 10,
    freshness: 'FRESH',
    bars: barsPullback(),
  })
  assert.ok(res.notes.includes(T_SIGNAL_REASON.CANNOT_SELL))
  assert.ok(!res.signals.some((s) => s.signal_type === T_SIGNAL_BUY_WATCH))
}

{
  const res = buildHoldingTSignal({
    hasPosition: true,
    canSell: true,
    costPrice: 10,
    freshness: 'STALE',
    bars: barsPullback(),
  })
  assert.equal(res.signals.length, 0)
  assert.ok(res.notes.includes(T_SIGNAL_REASON.PRICE_STALE))
}

{
  const res = buildHoldingTSignal({
    hasPosition: true,
    canSell: true,
    costPrice: 10,
    freshness: 'FRESH',
    bars: [{ close: 10, high: 10.1, low: 9.9 }],
  })
  assert.ok(res.notes.includes(T_SIGNAL_REASON.NO_5M_BARS))
}

{
  const res = buildHoldingTSignal({ hasPosition: false, bars: barsPullback(), costPrice: 10 })
  assert.ok(res.notes.includes(T_SIGNAL_REASON.NO_POSITION))
}

{
  const markers = holdingTSignalsToChartMarkers([
    { signal_type: T_SIGNAL_BUY_WATCH, signal_time: '10:35', signal_price: 10.03, reasons: ['NEAR_COST'] },
  ])
  assert.equal(markers[0].type, 'BUY')
  assert.equal(markers[0].time, '10:35')
}

console.log('holdingTSignal.test.mjs: ok')
